File: lib/stock/model_stock.dart
// lib/models/stock.dart
class ModelStock {
  DateTime date;
  String securityCode;
  String securityName;
  double? tradedNumber;
  int? transactionsNumber;
  double? transactionAmount;
  double? openingPrice;
  double? highestPrice;
  double? lowestPrice;
  double closingPrice;
  String? change;
  double? priceDifference;
  double? finalRevealBuyingPrice;
  double? finalRevealBuyingVolume;
  double? finalRevealSellingPrice;
  double? finalRevealSellingVolume;
  double? peRatio;
  String? source;
  double? ma5;
  double? ma20;
  double? high20;
  double? pctChange;
  double? vol5;
  double? rsi;
  bool? isRising;

  ModelStock(
      {required this.date,
      required this.securityCode,
      required this.securityName,
      this.tradedNumber,
      this.transactionsNumber,
      this.transactionAmount,
      this.openingPrice,
      this.highestPrice,
      this.lowestPrice,
      required this.closingPrice,
      this.change,
      this.priceDifference,
      this.finalRevealBuyingPrice,
      this.finalRevealBuyingVolume,
      this.finalRevealSellingPrice,
      this.finalRevealSellingVolume,
      this.peRatio,
      this.source,
      this.ma5,
      this.ma20,
      this.high20,
      this.pctChange,
      this.vol5,
      this.rsi,
      this.isRising});

  Map<String, dynamic> toJson() {
    return {
      'date': date.toIso8601String(),
      'security_code': securityCode,
      'security_name': securityName,
      'traded_number': tradedNumber,
      'transactions_number': transactionsNumber,
      'transaction_amount': transactionAmount,
      'opening_price': openingPrice,
      'highest_price': highestPrice,
      'lowest_price': lowestPrice,
      'closing_price': closingPrice,
      'change': change,
      'price_difference': priceDifference,
      'final_reveal_buying_price': finalRevealBuyingPrice,
      'final_reveal_buying_volume': finalRevealBuyingVolume,
      'final_reveal_selling_price': finalRevealSellingPrice,
      'final_reveal_selling_volume': finalRevealSellingVolume,
      'pe_ratio': peRatio,
      'source': source,
      'ma5': ma5,
      'ma20': ma20,
      'high20': high20,
      'pctChange': pctChange,
      'vol5': vol5,
      'rsi': rsi,
    };
  }

  factory ModelStock.fromJson(Map<String, dynamic> json) {
    return ModelStock(
      date: DateTime.parse(json['date']),
      securityCode: json['security_code'],
      securityName: json['security_name'],
      tradedNumber: (json['traded_number'] as num?)?.toDouble(),
      transactionsNumber: json['transactions_number'] as int?,
      transactionAmount: (json['transaction_amount'] as num?)?.toDouble(),
      openingPrice: (json['opening_price'] as num?)?.toDouble(),
      highestPrice: (json['highest_price'] as num?)?.toDouble(),
      lowestPrice: (json['lowest_price'] as num?)?.toDouble(),
      closingPrice: (json['closing_price'] as num?)?.toDouble() ?? 0,
      change: json['change']?.toString(),
      priceDifference: (json['price_difference'] as num?)?.toDouble(),
      finalRevealBuyingPrice:
          (json['final_reveal_buying_price'] as num?)?.toDouble(),
      finalRevealBuyingVolume:
          (json['final_reveal_buying_volume'] as num?)?.toDouble(),
      finalRevealSellingPrice:
          (json['final_reveal_selling_price'] as num?)?.toDouble(),
      finalRevealSellingVolume:
          (json['final_reveal_selling_volume'] as num?)?.toDouble(),
      peRatio: (json['pe_ratio'] as num?)?.toDouble(),
      source: json['source']?.toString(),
      ma5: (json['ma5'] as num?)?.toDouble(),
      ma20: (json['ma20'] as num?)?.toDouble(),
      high20: (json['high20'] as num?)?.toDouble(),
      pctChange: (json['pctChange'] as num?)?.toDouble(),
      vol5: (json['vol5'] as num?)?.toDouble(),
      rsi: (json['rsi'] as num?)?.toDouble(),
    );
  }
}

File: python_ai_service/stock/model_stock.py
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import (
    Float,
    Integer,
    String,
    Column,
    Text,
    TIMESTAMP,
)
from sqlalchemy.sql import func

Base = declarative_base()
MODEL_CACHE = {}

def create_stock_model(table_name: str):
    if table_name in MODEL_CACHE:
        return MODEL_CACHE[table_name]
    if table_name == "stock_daily_price":
        model = type(
            f"{table_name}_model",  # class 名稱
            (Base,),
            {
                "__tablename__": table_name,
                "__table_args__": {"extend_existing": True},  
                "security_code": Column(String, primary_key=True),
                "security_name": Column(String),
                "traded_number": Column(Float),
                "transactions_number": Column(Integer),
                "transaction_amount": Column(Float),
                "opening_price": Column(Float),
                "highest_price": Column(Float),
                "lowest_price": Column(Float),
                "closing_price": Column(Float),
                "change": Column(Text),
                "price_difference": Column(Float),
                "final_reveal_buying_price": Column(Float),
                "final_reveal_buying_volume": Column(Float),
                "final_reveal_selling_price": Column(Float),
                "final_reveal_selling_volume": Column(Float),
                "pe_ratio": Column(Float),
                "date": Column(TIMESTAMP(timezone=True), primary_key=True),
                "created_at": Column(TIMESTAMP(timezone=True), server_default=func.now()),
                "source": Column(Text),
                "ma5": Column(Float),
                "ma20": Column(Float),
                "high20": Column(Float),
                "pct_change": Column(Float),
                "vol5": Column(Float),
                "rsi": Column(Float),
            }
        )
    else:
        model = type(
            f"{table_name}_model",  # class 名稱
            (Base,),
            {
                "__tablename__": table_name,
                "__table_args__": {"extend_existing": True},  
                "date": Column(TIMESTAMP(timezone=True), primary_key=True),
                "type": Column(String, primary_key=True),
            }
        )
    MODEL_CACHE[table_name] = model
    return model